// @pothos/plugin-relay/lib/input-field-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
const _core = require("@pothos/core");
const inputFieldBuilder = _core.InputFieldBuilder.prototype;
inputFieldBuilder.globalIDList = function globalIDList(options = {}) {
    return this.idList({
        ...options,
        extensions: {
            ...options.extensions,
            isRelayGlobalID: true
        }
    });
};
inputFieldBuilder.globalID = function globalID(options = {}) {
    return this.id({
        ...options,
        extensions: {
            ...options.extensions,
            isRelayGlobalID: true
        }
    });
};
inputFieldBuilder.connectionArgs = function connectionArgs() {
    const { // TODO(breaking) make this default match other cursor fields
    cursorType ='ID' , beforeArgOptions ={} , afterArgOptions ={} , firstArgOptions ={} , lastArgOptions ={}  } = this.builder.options.relayOptions;
    return {
        before: this.field({
            ...beforeArgOptions,
            type: cursorType,
            required: false
        }),
        after: this.field({
            ...afterArgOptions,
            type: cursorType,
            required: false
        }),
        first: this.int({
            ...firstArgOptions,
            required: false
        }),
        last: this.int({
            ...lastArgOptions,
            required: false
        })
    };
};
