// @pothos/plugin-prisma/esm/field-builder.js
import { ObjectRef, RootFieldBuilder } from '@pothos/core';
import { ModelLoader } from './model-loader.js';
import { getCursorFormatter, getCursorParser, resolvePrismaCursorConnection } from './util/cursors.js';
import { getRefFromModel } from './util/datamodel.js';
import { queryFromInfo } from './util/map-query.js';
export * from './prisma-field-builder.js';
const fieldBuilderProto = RootFieldBuilder.prototype;
fieldBuilderProto.prismaField = function prismaField({ type, resolve, ...options }) {
    const modelOrRef = Array.isArray(type) ? type[0] : type;
    const typeRef = typeof modelOrRef === "string" ? getRefFromModel(modelOrRef, this.builder) : modelOrRef;
    const typeParam = Array.isArray(type) ? [
        typeRef
    ] : typeRef;
    return this.field({
        ...options,
        type: typeParam,
        resolve: (parent, args, context, info) => {
            const query = queryFromInfo({
                context,
                info
            });
            return resolve(query, parent, args, context, info);
        }
    });
};
fieldBuilderProto.prismaFieldWithInput = function prismaFieldWithInput({ type, resolve, ...options }) {
    const modelOrRef = Array.isArray(type) ? type[0] : type;
    const typeRef = typeof modelOrRef === "string" ? getRefFromModel(modelOrRef, this.builder) : modelOrRef;
    const typeParam = Array.isArray(type) ? [
        typeRef
    ] : typeRef;
    return this.fieldWithInput({
        ...options,
        type: typeParam,
        resolve: (parent, args, context, info) => {
            const query = queryFromInfo({
                context,
                info
            });
            return resolve(query, parent, args, context, info);
        }
    });
};
fieldBuilderProto.prismaConnection = function prismaConnection({ type, cursor, maxSize, defaultSize, resolve, totalCount, ...options }, connectionOptions = {}, edgeOptions = {}) {
    var _this_builder_configStore_getTypeConfig_extensions;
    const ref = typeof type === "string" ? getRefFromModel(type, this.builder) : type;
    const typeName = this.builder.configStore.getTypeConfig(ref).name;
    const model = (_this_builder_configStore_getTypeConfig_extensions = this.builder.configStore.getTypeConfig(ref).extensions) === null || _this_builder_configStore_getTypeConfig_extensions === void 0 ? void 0 : _this_builder_configStore_getTypeConfig_extensions.pothosPrismaModel;
    const formatCursor = getCursorFormatter(model, this.builder, cursor);
    const parseCursor = getCursorParser(model, this.builder, cursor);
    const cursorSelection = ModelLoader.getCursorSelection(ref, model, cursor, this.builder);
    const fieldRef = this.connection({
        ...options,
        type: ref,
        resolve: (parent, args, context, info) => resolvePrismaCursorConnection({
            query: queryFromInfo({
                context,
                info,
                select: cursorSelection,
                path: [
                    "edges",
                    "node"
                ],
                typeName
            }),
            ctx: context,
            parseCursor,
            maxSize,
            defaultSize,
            args,
            totalCount: totalCount && (() => totalCount(parent, args, context, info))
        }, formatCursor, (query) => resolve(query, parent, args, context, info))
    }, connectionOptions instanceof ObjectRef ? connectionOptions : {
        ...connectionOptions,
        fields: totalCount ? (t) => {
            var _connectionOptions_fields;
            return {
                totalCount: t.int({
                    nullable: false,
                    resolve: (parent, args, context) => {
                        var _parent_totalCount;
                        return (_parent_totalCount = parent.totalCount) === null || _parent_totalCount === void 0 ? void 0 : _parent_totalCount.call(parent);
                    }
                }),
                ...(_connectionOptions_fields = connectionOptions.fields) === null || _connectionOptions_fields === void 0 ? void 0 : _connectionOptions_fields.call(connectionOptions, t)
            };
        } : connectionOptions.fields,
        extensions: {
            ...connectionOptions === null || connectionOptions === void 0 ? void 0 : connectionOptions.extensions
        }
    }, edgeOptions);
    return fieldRef;
};
